client/gutenberg/editor/calypsoify-iframe.jsx

```diff
@@ -5,40 +5,136 @@
  */
 import React, { Component, Fragment } from 'react';
 import { connect } from 'react-redux';
-import { pickBy } from 'lodash';
+import { map, pickBy } from 'lodash';
 
 /**
  * Internal dependencies
  */
+import MediaLibrarySelectedData from 'components/data/media-library-selected-data';
+import MediaModal from 'post-editor/media-modal';
+import MediaActions from 'lib/media/actions';
 import { getSelectedSiteId } from 'state/ui/selectors';
-import { getSiteOption, getSiteAdminUrl } from 'state/sites/selectors';
+import { getSiteOption, getSiteAdminUrl, getSiteSlug } from 'state/sites/selectors';
 import { addQueryArgs } from 'lib/route';
+import {
+	getEnabledFilters,
+	getDisabledDataSources,
+	mediaCalypsoToGutenberg,
+} from './hooks/components/media-upload/utils';
 
 /**
  * Style dependencies
  */
 import './style.scss';
 
 class CalypsoifyIframe extends Component {
+	state = {
+		isMediaModalVisible: false,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.iframeRef = React.createRef();
+	}
+
+	componentDidMount() {
+		window.addEventListener( 'message', this.onMessage, false );
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener( 'message', this.onMessage, false );
+	}
+
+	onMessage = ( { data, origin } ) => {
+		if ( ! data || origin.indexOf( this.props.siteSlug ) < 0 ) {
+			return;
+		}
+
+		const { action, type } = data;
+
+		if ( 'gutenbergIframeMessage' !== type ) {
+			return;
+		}
+
+		if ( 'loaded' === action ) {
+			const { port1: portToIframe, port2: portForIframe } = new MessageChannel();
+
+			this.iframePort = portToIframe;
+			this.iframePort.addEventListener( 'message', this.oniFramePortMessage, false );
+			this.iframePort.start();
+
+			this.iframeRef.current.contentWindow.postMessage( { action: 'initPort' }, '*', [
+				portForIframe,
+			] );
+		}
+	};
+
+	oniFramePortMessage = ( { data } ) => {
+		const { action, payload } = data;
+
+		if ( 'openMediaModal' === action ) {
+			const { siteId } = this.props;
+			const { allowedTypes, gallery, multiple, value } = payload;
+
+			this.setState( { isMediaModalVisible: true, allowedTypes, gallery, multiple } );
+
+			if ( ! value ) {
+				MediaActions.setLibrarySelectedItems( siteId, [] );
```

Comment:
not sure if it matters here but we should remember the async nature of `setState()`. if we want to update our UI and issue media calls which could also affect that UI then we'll want to force sequencing on those actions, which we can do by passing a callback as the second parameter to `setState()`

```js
this.setState( { … }, () => {
	if ( ! value ) {
		return MediaActions.setLibrarySelectionItems( siteId, [] );
	}

	const selectedItems = Array.isArray( value )
		…

	MediaActions.setLibrarySelectedItems( siteId, selectedItems );
} );
```